---
title: "Subclonality of expression subtypes"
author: Ludwig Geistlinger, Sehyun Oh, Markus Riester, and Levi Waldron
output:
  BiocStyle::html_document:
    toc: true
    toc_depth: 2
vignette: >
  % \VignetteIndexEntry{Functional enrichment analysis of high-throughput omics data}
  % \VignetteEngine{knitr::rmarkdown}
---

```{r style, echo = FALSE}
suppressPackageStartupMessages({ 
    library(subtypeHeterogeneity) 
    library(consensusOV)
    library(RaggedExperiment)
    library(ComplexHeatmap)
    library(plotly)
})
```

# Setup
```{r setup}
library(subtypeHeterogeneity)
library(consensusOV)
library(RaggedExperiment)
library(ComplexHeatmap)
library(plotly)
```

```{r cols}
cb.pink <- "#CC79A7"
cb.red <- "#D55E00"
cb.blue <- "#0072B2"
cb.yellow <- "#F0E442"
cb.green <- "#009E73"
cb.lightblue <- "#56B4E9"
cb.orange <- "#E69F00"

stcols <- c(cb.lightblue, cb.green, cb.orange, cb.pink) 
names(stcols) <- c("PRO", "MES", "DIF", "IMR")
```

# Data sources

## Cancer types
```{r ctypes}
RTCGAToolbox::getFirehoseDatasets()
```

## ABSOLUTE
```{r absolute}
data.dir <- system.file("extdata", package="subtypeHeterogeneity") 
absFile <- file.path(data.dir, "ABSOLUTE_grangeslist.rds")
absGRL <- readRDS(absFile) 
absGRL
```

## GISTIC2
```{r gistic}
gisticOV <- gistic2RSE(ctype="OV", peak="wide")
gisticOV
rowRanges(gisticOV)
assay(gisticOV)[1:5,1:5]
```

## Expression-based subtypes
### Broad subtypes
```{r broadSubtys}
ovsubs <- getBroadSubtypes(ctype="OV", clust.alg="CNMF")
dim(ovsubs)
head(ovsubs)
table(ovsubs[,"cluster"])
```

### OV subtypes from different studies
```{r pooledSubtys}
pooled.file <- file.path(data.dir, "pooled_subtypes.rds")
pooled.subs <- readRDS(pooled.file)
table(pooled.subs[,"data.source"])
table(pooled.subs[,"Verhaak"])  
```

### TCGA subtype consistency
```{r subtyCons}
tab <- mapOVSubtypes(ovsubs, pooled.subs)
tab
(ind <- sort( apply(tab, 1, which.max) ))
```

```{r mapSubs}
sts <- names(ind)[ovsubs[,"cluster"]]
ovsubs <- data.frame(ovsubs, subtype=sts, stringsAsFactors=FALSE)
```

### Subtype purity & ploidy
```{r subtyPP}
pp.file <- file.path(data.dir, "ABSOLUTE_Purity_Ploidy.rds")
puri.ploi <- readRDS(pp.file)
head(puri.ploi)
```

```{r plotPP, fig.width=8, fig.height=8}
plotSubtypePurityPloidy(ovsubs, puri.ploi)
```

```{r stratPur}
sebin <- stratifyByPurity(ovsubs, puri.ploi, method="equal.bin")
lengths(sebin)
squint <- stratifyByPurity(ovsubs, puri.ploi, method="quintile")
lengths(squint)
```

```{r plotPurStrat, fig.width=5, fig.height=5}
plotPurityStrata(ovsubs, puri.ploi)
```

# Test: gistic-subtype association
```{r subtyAssoc}
pvals <- testSubtypes(gisticOV, ovsubs, padj.method="none")
adj.pvals <- p.adjust(pvals, method="BH")
length(adj.pvals)
head(adj.pvals)
sum(adj.pvals < 0.1)
```

```{r plotAssoc, fig.width=5, fig.height=5}
hist(pvals, breaks=25, col="firebrick",
        xlab="Subtype association p-value", main="")
```

## Genomic distribution of subtype-associated CNAs 

```{r tcgaOvCnvGenes}
cnv.genes <- getCnvGenesFromTCGA()
```

```{r circosSubAssoc, fig.width=10, fig.height=10}
sig.ind <- adj.pvals < 0.1
mcols(gisticOV)$subtype <- testSubtypes(gisticOV, ovsubs, what="subtype")
mcols(gisticOV)$significance <- ifelse(sig.ind, "*", "")
circosSubtypeAssociation(gisticOV, cnv.genes)
```

```{r cnaPerSubtype, fig.width=5, fig.height=5}
plotNrCNAsPerSubtype(mcols(gisticOV)$type[sig.ind], mcols(gisticOV)$subtype[sig.ind])
```

# Subclonality

## Match subtyped samples and ABSOLUTE samples 
```{r matchCalls}
raOV <- getMatchedAbsoluteCalls(absGRL, ovsubs)
raOV
rowRanges(raOV)
```

## Summarize subclonality of ABSOLUTE calls in GISTIC regions 
```{r querySubcl}
subcl.gisticOV <- querySubclonality(raOV, query=rowRanges(gisticOV), sum.method="any")
dim(subcl.gisticOV)
subcl.gisticOV[1:5,1:5]
```

## Subclonality score
Def: Fraction of samples in which a mutation is subclonal.
```{r subclScore}
subcl.score <- rowMeans(subcl.gisticOV, na.rm=TRUE)
summary(subcl.score)
```

## Correlation: subtype-association score and subclonality score
```{r cor}
assoc.score <- testSubtypes(gisticOV, ovsubs, what="statistic")
cor(assoc.score, subcl.score, method="spearman")
cor.test(assoc.score, subcl.score, method="spearman", exact=FALSE)$p.value
#corPermTest(gisticOV, ovsubs, subcl.score)
```

```{r plotCor, fig.width=6, fig.height=6}
plotCorrelation(assoc.score, subcl.score) 
```

### Purity-stratified analysis
```{r purStrat}
sapply(squint, function(ids) analyzeStrata(ids, absGRL, gisticOV, ovsubs))
sapply(sebin[-1], function(ids) analyzeStrata(ids, absGRL, gisticOV, ovsubs))
```

### Only copy losses and single copy gains
```{r corNoHighAmp}
noHighAmp <- getMatchedAbsoluteCalls(absGRL, ovsubs, max.cn=3)
subcl.noHighAmp <- querySubclonality(noHighAmp, query=rowRanges(gisticOV))
subcl.noHighAmp <- rowMeans(subcl.noHighAmp, na.rm=TRUE)
cor(assoc.score, subcl.noHighAmp, method="spearman")
cor.test(assoc.score, subcl.noHighAmp, method="spearman", exact=FALSE)$p.value
#corPermTest(gisticOV, ovsubs, subcl.noHighAmp)
```

```{r plotCorNHA, fig.width=6, fig.height=6}
plotCorrelation(assoc.score, subcl.noHighAmp)
```

### Extend peaks by 500 kb up- and downstream 
```{r cor5kb, fig.width=6, fig.height=6}
subcl.gisticOV.5kb <- querySubclonality(raOV, query=rowRanges(gisticOV), ext=500000)
subcl.score <- rowMeans(subcl.gisticOV.5kb, na.rm=TRUE)
summary(subcl.score)
cor(assoc.score, subcl.score, method="spearman")
cor.test(assoc.score, subcl.score, method="spearman", exact=FALSE)$p.value
plotCorrelation(assoc.score, subcl.score, 
    subtypes=mcols(gisticOV)$subtype, stcols=stcols)
```

```{r annoOVranges}
rowRanges(gisticOV)$adj.pval <- adj.pvals
rowRanges(gisticOV)$subcl <- subcl.score 
```

## Assessing subclonality calls with PureCN

### Read and order
```{r pureCN.pp}
ppp.file <- file.path(data.dir, "PureCN_Purity_Ploidy.txt")
puriploi.pcn <- read.table(ppp.file) 
puriploi.pcn <- puriploi.pcn[!duplicated(puriploi.pcn[,1]),]
rownames(puriploi.pcn) <- puriploi.pcn[,"Sampleid"]
puriploi.pcn <- puriploi.pcn[,-1]
puriploi.pcn <- as.matrix(puriploi.pcn)
absdiff <- abs(puriploi.pcn[,c("purity", "ploidy")] - 
                puriploi.pcn[,c("Purity_wes", "Ploidy_wes")])
ind <- do.call(order, as.data.frame(absdiff))
puriploi.pcn <- puriploi.pcn[ind,]
head(puriploi.pcn)
cor(puriploi.pcn[,"purity"], puriploi.pcn[,"Purity_wes"], use="complete.obs")
cor(puriploi.pcn[,"Purity_wes"], puriploi.pcn[,"Purity_snp6"], use="complete.obs")
```

### CN concordance with ABSOLUTE
PureCN calls for S0293689 capture kit: 
```{r pureCN.calls}
pcn.call.file <- file.path(data.dir, "PureCN_OV_S0293689_grangeslist.rds")
pcn.calls <- readRDS(pcn.call.file)
pcn.calls
# select samples that intersect with ABSOLUTE data
isect.samples <- intersect(names(pcn.calls), names(absGRL))
pcn.calls <- pcn.calls[isect.samples]
pcn.calls.ra <- RaggedExperiment(pcn.calls) 
pcn.calls.ra
```

Get corresponding ABSOLUTE calls:
```{r isect.abs.calls}
abs.calls <- absGRL[isect.samples]
abs.calls <- as.data.frame(abs.calls)[,c(2:5,8:10)]
totalCN <- abs.calls[,"Modal_HSCN_1"] + abs.calls[,"Modal_HSCN_2"]
abs.calls <- data.frame(abs.calls[,1:4], CN=totalCN, score=abs.calls[,"score"])
abs.calls <- makeGRangesListFromDataFrame(abs.calls, 
    split.field="group_name", keep.extra.columns=TRUE)
abs.calls
abs.calls.ra <- RaggedExperiment(abs.calls)
abs.calls.ra <- abs.calls.ra[,colnames(pcn.calls.ra)]
```

OVC GISTIC2 regions on hg19 (ABSOLUTE) and hg38 (PureCN)
```{r gisticHG38}
gisticOV.ranges.hg19 <- file.path(data.dir, "gisticOV_ranges_hg19.txt") 
gisticOV.ranges.hg19 <- scan(gisticOV.ranges.hg19, what="character")
gisticOV.ranges.hg19 <- GRanges(gisticOV.ranges.hg19) 
gisticOV.ranges.hg19 
gisticOV.ranges.hg38 <- file.path(data.dir, "gisticOV_ranges_hg38.txt") 
gisticOV.ranges.hg38 <- scan(gisticOV.ranges.hg38, what="character")
gisticOV.ranges.hg38 <- GRanges(gisticOV.ranges.hg38) 
gisticOV.ranges.hg38
```

Summarize calls in GISTIC regions using either  
 
- the largest call overlapping a GISTIC region  
- a weighted mean of calls overlapping a GISTIC region 
```{r qred}
.largest <- function(score, range, qrange) 
{
    return.type <- class(score[[1]])
    default.value <- do.call(return.type, list(1))
    ind <- which.max(width(range))
    res <- vapply(seq_along(score), 
           function(i) score[[i]][ind[i]], default.value)
    return(res)
}
pcn.rassay <- qreduceAssay(pcn.calls.ra, query=gisticOV.ranges.hg38, 
                    simplifyReduce=.largest, background=2)
abs.rassay <- qreduceAssay(abs.calls.ra, query=gisticOV.ranges.hg19,
                    simplifyReduce=.largest, background=2)


.weightedmean <- function(score, range, qrange)
{
    w <- width(range)
    s <- sum(score * w) / sum(w)
    return(round(s))
}

pcn.rassay2 <- qreduceAssay(pcn.calls.ra, query=gisticOV.ranges.hg38,
                    simplifyReduce=.weightedmean, background=2)
abs.rassay2 <- qreduceAssay(abs.calls.ra, query=gisticOV.ranges.hg19,
                    simplifyReduce=.weightedmean, background=2)
```

Evaluate per-sample concordance:

- fraction of GISTIC regions with equal copy number  
- fraction of GISTIC regions agreeing in alteration type (amplification / deletion)  
- fraction of GISTIC regions that deviate by max. 1 copy

```{r evalConcord}
# largest
fract.equal <- vapply(seq_along(isect.samples), 
        function(i) mean(pcn.rassay[,i] == abs.rassay[,i]), 
        numeric(1))
fract.type <- vapply(seq_along(isect.samples), 
        function(i) mean(((pcn.rassay[,i] < 2) & (abs.rassay[,i] < 2)) |
                          ((pcn.rassay[,i] > 2) & (abs.rassay[,i] > 2)) |  
                          ((pcn.rassay[,i] == 2) & (abs.rassay[,i] == 2))), 
        numeric(1))
fract.diff1 <- vapply(seq_along(isect.samples), 
        function(i) mean((pcn.rassay[,i] == abs.rassay[,i]) | 
                            (pcn.rassay[,i] == abs.rassay[,i] + 1) |
                            (pcn.rassay[,i] == abs.rassay[,i] - 1)), 
        numeric(1))

# weighted mean
fract.equal2 <- vapply(seq_along(isect.samples), 
        function(i) mean(pcn.rassay2[,i] == abs.rassay2[,i]), 
        numeric(1))
fract.type2 <- vapply(seq_along(isect.samples), 
        function(i) mean(((pcn.rassay2[,i] < 2) & (abs.rassay2[,i] < 2)) |
                          ((pcn.rassay2[,i] > 2) & (abs.rassay2[,i] > 2)) |  
                          ((pcn.rassay2[,i] == 2) & (abs.rassay2[,i] == 2))), 
        numeric(1))
fract.diff12 <- vapply(seq_along(isect.samples), 
        function(i) mean((pcn.rassay2[,i] == abs.rassay2[,i]) | 
                            (pcn.rassay2[,i] == abs.rassay2[,i] + 1) |
                            (pcn.rassay2[,i] == abs.rassay2[,i] - 1)), 
        numeric(1))
```

Plot concordance:
```{r plotConcord}
par(mar=c(5.1, 4.1, 4.1, 8.1), xpd=TRUE)
boxplot(fract.equal, fract.equal2, 
        fract.type, fract.type2,
        fract.diff1, fract.diff12, col=rep(c(cb.blue, cb.red), 3),
        names=rep(c("equal", "type", "diff1"), each=2),
        ylab="Fraction of concordant GISTIC2 regions")
legend("topright", inset=c(-0.3,0), legend=c("largest", "wmean"),
        col=c(cb.blue, cb.red), lwd=2)
```


### Subclonality concordance with ABSOLUTE 

Summarize subclonality of GISTIC2 regions

```{r subclConcord}
# PureCN
maxScore <- function(score, range, qrange) max(score, na.rm=TRUE)
pcn.rassay.sc <- RaggedExperiment::qreduceAssay(pcn.calls.ra, 
                query=gisticOV.ranges.hg38, 
                simplifyReduce=maxScore, i="score", background=NA)
pcn.subcl <- rowMeans(pcn.rassay.sc, na.rm=TRUE)
summary(pcn.subcl)

#ABSOLUTE
abs.rassay.sc <- RaggedExperiment::qreduceAssay(abs.calls.ra, 
                query=gisticOV.ranges.hg19, 
                simplifyReduce=maxScore, i="score", background=NA)
abs.subcl <- rowMeans(abs.rassay.sc, na.rm=TRUE)
summary(abs.subcl)
```

Correlation 
```{r corConcord}
cor.test(pcn.subcl, abs.subcl, method="spearman")
```


```{r selRegions}
sort(abs.subcl[pcn.subcl > 0.4], decreasing=TRUE)
```

### Select large losses called subclonal by ABSOLUTE 
```{r selectAbsCalls}
selectAbsCalls <- function(sample, min.size=3000000)
{
    gr <- absGRL[[sample]]
    gr.3MB <- width(gr) > min.size
    is.loss <- (gr$Modal_HSCN_1 + gr$Modal_HSCN_2) < 2 
    is.subcl <- gr$score == 1

    ind <- gr.3MB & is.loss & is.subcl
    return(gr[ind])
}
(abs.calls <- selectAbsCalls(rownames(puriploi.pcn)[1]))
```

### liftOver hg19-based ABSOLUTE calls to compare with hg38-based PureCN calls 
- UCSC online liftOver (https://genome.ucsc.edu/cgi-bin/hgLiftOver)
- does smoothing and filtering on top of the chain mappings
- `rtracklayer::liftOver` only provides chain mappings

```{r liftOver}
chain.file <- file.path(data.dir, "hg19ToHg38.over.chain")
ch <- rtracklayer::import.chain(chain.file)
(hg19.call <- absGRL[[1]][1])
(blocks <- rtracklayer::liftOver(hg19.call, ch))
```

```{r liftOverFilterSmooth}
blocks <- unlist(blocks)
# harmonize chromosome by majority vote
tab <- table(as.character(seqnames(blocks))) 
nmax <- names(tab)[which.max(tab)]
# collapse blocks
blocks <- blocks[seqnames(blocks) == nmax]
(hg38.call <- range(reduce(blocks)))
```

### Select GISTIC regions of high subtype association and subclonality
```{r selecttopTables}
tests <- testSubtypes(gisticOV, ovsubs, what="full")
```

```{r selecttop}
# highest subtype association
ind <- order(assoc.score, decreasing=TRUE)
rowRanges(gisticOV)[ind[1:2]]
x <- tests[[ind[1]]]
x$observed
x$expected
diff <- x$observed - x$expected
(csums <- colSums(diff^2 / x$expected))
x <- tests[[ind[2]]]
x$observed
x$expected
diff <- x$observed - x$expected
(csums <- colSums(diff^2 / x$expected))

# highest subclonality
ind <- order(subcl.score, decreasing=TRUE)
rowRanges(gisticOV)[ind[1:2]]
x <- tests[[ind[1]]]
x$observed
x$expected
diff <- x$observed - x$expected
(csums <- colSums(diff^2 / x$expected))
x <- tests[[ind[2]]]
x$observed
x$expected
diff <- x$observed - x$expected
(csums <- colSums(diff^2 / x$expected))
```

Plot regions:
```{r intersectSamplesABSOLUTEandSubtype}
isamples <- intersect(rownames(ovsubs), names(absGRL))
absov <- absGRL[isamples]
```

```{r plotHighSubcl, fig.width=6, fig.height=6}
# myc
r <- GRanges("chr8:128694547-129309687")
g <- GRanges(c("chr8:128747680-128753674", "chr8:128806779-129113499"))
names(g) <- c("MYC", "PVT1")
genome(g) <- genome(r) <- "hg19" 

ov.olaps <- subsetByOverlaps(unlist(absov), r)
#as a GRangesList: splitAsList(olaps, names(olaps))
ov.olaps$State <- ovsubs[names(ov.olaps), "cluster"]
gvizRegion(r, g, ov.olaps)
```


# Other cancer types

```{r allCTypes}
allFile <- file.path(data.dir, "all_cancer_types.rds")
allCT <- readRDS(allFile)
```

## Number of samples
```{r plotNrSamples, fig.width=12, fig.height=6}
st <- sapply(allCT, function(ct) nrow(ct$subtypes))
gs <- sapply(allCT, function(ct) ncol(ct$gistic))
as <- sapply(allCT, function(ct) sum(rownames(ct$subtypes) %in% names(absGRL)))

plotNrSamples(gs, st, as) 
```

## Number of subtypes & GISTIC regions
```{r descrStats, fig.width=6, fig.height=6}
nr.sts <- sapply(allCT, function(ct) length(unique(ct$subtypes[,1])))
nr.gregs <- sapply(allCT, function(ct) nrow(ct$gistic))
plotNrSubtypesVsNrGisticRegions(nr.sts, nr.gregs)
```

## Subclonality
```{r subclDistrs, fig.width=12, fig.height=6}
subcl.scores <- lapply(allCT, function(ct) ct$subcl.score)
plotSubclonalityDistributions(subcl.scores)
```

## Correlation: subtype association & subclonality
```{r corVolcano, fig.width=9, fig.height=6}
rho <- sapply(allCT, function(ct) ct$rho)
p <- sapply(allCT, function(ct) ct$p)
volcanoCorrelation(rho, p)
```

## Intrinsic subtypes: sarcoma
```{r sarcSubtyAssoc}
gisticSARC <- gistic2RSE(ctype="SARC", peak="wide")
sarcsubs <- getBroadSubtypes(ctype="SARC", clust.alg="CNMF")
table(sarcsubs$cluster)
pvals <- testSubtypes(gisticSARC, sarcsubs, padj.method="none")
length(pvals)
adj.pvals <- p.adjust(pvals, method="BH")
sum(adj.pvals < 0.1)
hist(pvals, breaks=25, col="firebrick",xlab="Subtype association p-value", main="")
```


```{r sarcClinical}
sarc.clin <- RTCGAToolbox::getFirehoseData(dataset="SARC", runDate="20160128")
sarc.clin <- RTCGAToolbox::getData(sarc.clin, "clinical")
ids <- rownames(sarcsubs)
ids <- tolower(ids)
ids <- gsub("-", ".", ids)
ids <- substring(ids, 1, 12)
sarcsubs$histType <- sarc.clin[ids,"histological_type"] 
lapply(1:3, function(cl) table(sarcsubs[sarcsubs$cluster == cl,"histType"]))
```

```{r sarc}
assoc.score <- testSubtypes(gisticSARC, sarcsubs, what="statistic")
raSARC <- getMatchedAbsoluteCalls(absGRL, sarcsubs)
subcl.gisticSARC <- querySubclonality(raSARC, query=rowRanges(gisticSARC), sum.method="any")
subcl.score <- rowMeans(subcl.gisticSARC, na.rm=TRUE)
summary(subcl.score)
cor(assoc.score, subcl.score, method="spearman")
cor.test(assoc.score, subcl.score, method="spearman", exact=FALSE)$p.value
```

```{r sarcPlot, fig.width=6, fig.height=6}
sarc.cols <- stcols[c(1,3,2)]
names(sarc.cols) <- c("MFS/UPS", "LMS", "DDLPS")
sts <- testSubtypes(gisticSARC, sarcsubs, what="subtype")
sts <- names(sarc.cols)[sts]
plotCorrelation(assoc.score, subcl.score, subtypes=sts, stcols=sarc.cols)
```

### Overlapping SCNAs between OV and SARC
```{r ovsarcranges}
sarcranges <- rowRanges(gisticSARC)
sarcranges$adj.pval <- adj.pvals
sarcranges$subcl <- subcl.score
ovranges <- rowRanges(gisticOV)
plotCommonSCNAs(ovranges, sarcranges)
```

Plot regions:
```{r intersectSamplesABSOLUTEandSubtypeSARC}
isamples <- intersect(rownames(sarcsubs), names(absGRL))
abssarc <- absGRL[isamples]
```

```{r plotHighSubclSarc}
# myc
sarc.olaps <- subsetByOverlaps(unlist(abssarc), r)
#as a GRangesList: splitAsList(olaps, names(olaps))
sarc.olaps$State <- sarcsubs[names(sarc.olaps), "cluster"]
gvizRegion(r, g, ov.olaps, sarc.olaps)
```


<!--
## Intrinsic subtypes: breast cancer
Subtypes: luminal A, luminal B, basal-like, and HER2-enriched
```{r brcaSubtyAssoc}
gisticBRCA <- gistic2RSE(ctype="BRCA", peak="wide")
brcasubs <- getBroadSubtypes(ctype="BRCA", clust.alg="CNMF")
table(brcasubs$cluster)
pvals <- testSubtypes(gisticBRCA, brcasubs, padj.method="none")
length(pvals)
adj.pvals <- p.adjust(pvals, method="BH")
sum(adj.pvals < 0.1)
hist(pvals, breaks=25, col="firebrick",xlab="Subtype association p-value", main="")
```

```{r brca}
brca <- allCT$BRCA
names(brca)
brca[1:2]
```

## Late subtypes: glioblastoma
Subtypes: mesenchymal, proneural, neural, and classical
```{r gbmSubtyAssoc}
gisticGBM <- gistic2RSE(ctype="GBM", peak="wide")
gbmsubs <- getBroadSubtypes(ctype="GBM", clust.alg="CNMF")
table(gbmsubs$cluster)
pvals <- testSubtypes(gisticGBM, gbmsubs, padj.method="none")
length(pvals)
adj.pvals <- p.adjust(pvals, method="BH")
sum(adj.pvals < 0.1)
hist(pvals, breaks=25, col="firebrick",xlab="Subtype association p-value", main="")
```

```{r gbm}
gbm <- allCT$GBM
gbm[1:2]
```
-->

# Subtype classification in single-cell RNA-seq data 

## Exploratory analysis
```{r scRNAseq}
sc.file <- file.path(data.dir, "scRNAseq_logTPM.rds")
sc <- readRDS(sc.file)
sc[1:5, 1:5]
```

```{r explore.scRNAseq}
library(MASS)
data(GSE14764.eset)
boxplot(exprs(GSE14764.eset))
boxplot(sc)
ma.vals <- as.vector(exprs(GSE14764.eset))
sc.vals <- as.vector(sc)
truehist(ma.vals, nbins=50) 
truehist(sc.vals, nbins=50)
```

```{r exploreZeroInflat}
sc0 <- sc
for(i in seq_len(nrow(sc)))
    for(j in seq_len(ncol(sc)))
        if(is.na(sc[i,j])) sc0[i,j] <- 0
MASS::truehist(as.vector(sc0), nbins=50)
perc0 <- apply(sc0, 2, function(x) mean(x==0) * 100)
barplot(perc0,ylab="%genes with log2TPM = 0")
```

## Subtype classification

Genes considered by the `consensusOV` classifier:
```{r consensusGenes}
library(consensusOV)
fnames <- lapply(
    consensusOV:::esets.rescaled.classified.filteredgenes, 
    Biobase::featureNames)
clgenes <- sort(unique(unlist(fnames)))
clgenes <- sub("^geneid.", "", clgenes)
clgenes
```

Single cell data restricted to classifier genes:
```{r scRestricted}
sc.file <- file.path(data.dir, "scRNAseq_consensusOV_genes.txt")
sc.expr <- as.matrix(read.delim(sc.file, row.names=1L))
dim(sc.expr)
sc.expr[1:5,1:5]
# exclude all zero 
ind <- apply(sc.expr, 1, function(x) all(x == 0))
sc.expr <- sc.expr[!ind,]
dim(sc.expr)
```

Get consensus subtypes:
```{r consensusSubtypes}
consensus.subtypes <- get.consensus.subtypes(sc.expr, rownames(sc.expr))
    # remove.using.cutoff = TRUE, percentage.dataset.removed = 0.75
```

Cell type annotation:
```{r cellAnnotation}
c2t.file <- file.path(data.dir, "scRNAseq_celltypes.txt")
c2t <- read.delim(c2t.file, as.is=TRUE)
n <- paste0("X", c2t[,1])
c2t <- c2t[,2]
names(c2t) <- n
head(c2t) 
table(c2t)
```

Expression heatmap incl. subtype and cell type annotation:
```{r scHeatmap}
library(ComplexHeatmap)
sind <- 2:ncol(sc.expr)

# subtype
st <- sub("_consensus$", "", consensus.subtypes$consensusOV.subtypes[sind])
st <- as.factor(st)

# cell type
ccol <- c("#B62A84", "#2AB68C")
names(ccol) <- unique(unname(c2t))
ct <- as.factor(c2t[colnames(sc.expr)[sind]])

scol <- list(Subtype=stcols, CellType=ccol)
df <- data.frame(Subtype = st, CellType=ct)
ha <- HeatmapAnnotation(df = df, col=scol)

expr <- log(sc.expr[,sind] + 1, base=2)

Heatmap(expr, name="log2TPM", top_annotation = ha,
    show_row_names=FALSE, show_column_names=FALSE,
    column_title="Cells", row_title="Genes")
```

Subtype by cell type:
```{r stByCT}
ept <- table(st[ct=="Epithelial"])
ept <- round(ept / sum(ept), digits=3)
ept
stro <- table(st[ct=="Stromal"])
stro <- round(stro / sum(stro), digits=3)
stro
```

Pie chart of subtype classification:
```{r plotlyPie, eval=FALSE}
ept <- ept[ept > 0]
cind <- match(names(ept), names(stcols))
plotlyPie(labels=names(ept), values=ept, colors=stcols[cind])
```

Display classification probabilites:
```{r rfprobHeatmap}
dat <- consensus.subtypes$rf.probs[sind,]
colnames(dat) <- sub("_consensus$", "", colnames(dat))
Heatmap(dat, show_row_names=FALSE, row_title="Cells", name="Prob")
```

Consider only top50% confident classifications: 
```{r top50}
subtr <- apply(consensus.subtypes$rf.probs[sind,], 1, function(row) sort(row)[3])
my.predictions.margins <- rowMax(consensus.subtypes$rf.probs[sind,]) - subtr
ecdf.evaluated <- ecdf(my.predictions.margins)(my.predictions.margins)
na.ind <- ecdf.evaluated < 0.5
my.predictions.margins[na.ind] <- NA
top50 <- !is.na(my.predictions.margins)
```

Max prob (subtype call) and margin score (diff to second highest prob) distribution:
```{r marginDist}
maxp <- rowMax(consensus.subtypes$rf.probs[sind,])[top50]
marg <- my.predictions.margins[top50]
boxplot(maxp, marg, names=c("max", "margin"), ylab="class.prob")
```

```{r stByCT-top50}
ept <- table(st[top50][ct[top50]=="Epithelial"])
ept <- round(ept / sum(ept), digits=3)
ept
stro <- table(st[top50][ct[top50]=="Stroma"])
stro <- round(stro / sum(stro), digits=3)
stro
```

Repeat analysis for logTPM > 3 (UofW single-cell threshold)
```{r exclLowExpr, eval=FALSE}
ind <- rowSums(expr) > 3
consensus.subtypes <- get.consensus.subtypes(
    sc.expr[ind,], 
    rownames(sc.expr)[ind])
```
