---
title: "Subclonality of expression subtypes"
author: Ludwig Geistlinger, Sehyun Oh, Markus Riester, and Levi Waldron
output:
  BiocStyle::html_document:
    toc: true
    toc_depth: 2
vignette: >
  % \VignetteIndexEntry{Functional enrichment analysis of high-throughput omics data}
  % \VignetteEngine{knitr::rmarkdown}
---

```{r style, echo = FALSE}
suppressPackageStartupMessages({ 
    library(subtypeHeterogeneity) 
    library(consensusOV)
})
```

# Setup
```{r setup}
library(subtypeHeterogeneity)
library(consensusOV)
```

```{r cols}
stcols <- c("steelblue", "darkseagreen3", "coral", "firebrick")
names(stcols) <- c("PRO", "MES", "DIF", "IMR")
```

# Data sources

## Cancer types
```{r ctypes}
RTCGAToolbox::getFirehoseDatasets()
```

## ABSOLUTE
```{r absolute}
data.dir <- system.file("extdata", package="subtypeHeterogeneity") 
absFile <- file.path(data.dir, "ABSOLUTE_grangeslist.rds")
absGRL <- readRDS(absFile) 
absGRL
```

## GISTIC2
```{r gistic}
gisticOV <- gistic2RSE(ctype="OV", peak="wide")
gisticOV
rowRanges(gisticOV)
assay(gisticOV)[1:5,1:5]
```

## Expression-based subtypes
### Broad subtypes
```{r broadSubtys}
ovsubs <- getBroadSubtypes(ctype="OV", clust.alg="CNMF")
dim(ovsubs)
head(ovsubs)
table(ovsubs[,"cluster"])
```

### OV subtypes from different studies
```{r pooledSubtys}
pooled.file <- file.path(data.dir, "pooled_subtypes.rds")
pooled.subs <- readRDS(pooled.file)
table(pooled.subs[,"data.source"])
table(pooled.subs[,"Verhaak"])  
```

### TCGA subtype consistency
```{r subtyCons}
tab <- mapOVSubtypes(ovsubs, pooled.subs)
tab
(ind <- sort( apply(tab, 1, which.max) ))
```

```{r mapSubs}
sts <- names(ind)[ovsubs[,"cluster"]]
ovsubs <- data.frame(ovsubs, subtype=sts, stringsAsFactors=FALSE)
```

### Subtype purity & ploidy
```{r subtyPP}
pp.file <- file.path(data.dir, "ABSOLUTE_Purity_Ploidy.rds")
puri.ploi <- readRDS(pp.file)
head(puri.ploi)
```

```{r plotPP, fig.width=8, fig.height=8}
plotSubtypePurityPloidy(ovsubs, puri.ploi)
```

```{r stratPur}
sebin <- stratifyByPurity(ovsubs, puri.ploi, method="equal.bin")
lengths(sebin)
squint <- stratifyByPurity(ovsubs, puri.ploi, method="quintile")
lengths(squint)
```

```{r plotPurStrat, fig.width=5, fig.height=5}
plotPurityStrata(ovsubs, puri.ploi)
```

# Test: gistic-subtype association
```{r subtyAssoc}
pvals <- testSubtypes(gisticOV, ovsubs, padj.method="none")
adj.pvals <- p.adjust(pvals, method="BH")
length(adj.pvals)
head(adj.pvals)
sum(adj.pvals < 0.1)
```

```{r plotAssoc, fig.width=5, fig.height=5}
hist(pvals, breaks=25, col="firebrick",
        xlab="Subtype association p-value", main="")
```

## Genomic distribution of subtype-associated CNAs 

```{r tcgaOvCnvGenes}
cnv.genes <- getCnvGenesFromTCGA()
```

```{r circosSubAssoc, fig.width=10, fig.height=10}
mcols(gisticOV)$subtype <- testSubtypes(gisticOV, ovsubs, what="subtype")
mcols(gisticOV)$significance <- ifelse(adj.pvals < 0.1, "*", "")
circosSubtypeAssociation(gisticOV, cnv.genes)
```

```{r cnaPerSubtype, fig.width=5, fig.height=5}
plotNrCNAsPerSubtype(mcols(gisticOV)$type, mcols(gisticOV)$subtype)
```

# Subclonality

## Match subtyped samples and ABSOLUTE samples 
```{r matchCalls}
raOV <- getMatchedAbsoluteCalls(absGRL, ovsubs)
raOV
rowRanges(raOV)
```

## Summarize subclonality of ABSOLUTE calls in GISTIC regions 
```{r querySubcl}
subcl.gisticOV <- querySubclonality(raOV, query=rowRanges(gisticOV), sum.method="any")
dim(subcl.gisticOV)
subcl.gisticOV[1:5,1:5]
```

## Subclonality score
Def: Fraction of samples in which a mutation is subclonal.
```{r subclScore}
subcl.score <- rowMeans(subcl.gisticOV, na.rm=TRUE)
summary(subcl.score)
```

## Correlation: subtype-association score and subclonality score
```{r cor}
assoc.score <- testSubtypes(gisticOV, ovsubs, what="statistic")
cor(assoc.score, subcl.score, method="spearman")
cor.test(assoc.score, subcl.score, method="spearman", exact=FALSE)$p.value
#corPermTest(gisticOV, ovsubs, subcl.score)
```

```{r plotCor, fig.width=6, fig.height=6}
plotCorrelation(assoc.score, subcl.score) 
```

### Purity-stratified analysis
```{r purStrat}
sapply(squint, function(ids) analyzeStrata(ids, absGRL, gisticOV, ovsubs))
sapply(sebin[-1], function(ids) analyzeStrata(ids, absGRL, gisticOV, ovsubs))
```

### Only copy losses and single copy gains
```{r corNoHighAmp}
noHighAmp <- getMatchedAbsoluteCalls(absGRL, ovsubs, max.cn=3)
subcl.noHighAmp <- querySubclonality(noHighAmp, query=rowRanges(gisticOV))
subcl.noHighAmp <- rowMeans(subcl.noHighAmp, na.rm=TRUE)
cor(assoc.score, subcl.noHighAmp, method="spearman")
cor.test(assoc.score, subcl.noHighAmp, method="spearman", exact=FALSE)$p.value
#corPermTest(gisticOV, ovsubs, subcl.noHighAmp)
```

```{r plotCorNHA, fig.width=6, fig.height=6}
plotCorrelation(assoc.score, subcl.noHighAmp)
```

### Extend peaks by 500 kb up- and downstream 
```{r cor5kb, fig.width=6, fig.height=6}
subcl.gisticOV.5kb <- querySubclonality(raOV, query=rowRanges(gisticOV), ext=500000)
subcl.score <- rowMeans(subcl.gisticOV.5kb, na.rm=TRUE)
summary(subcl.score)
cor(assoc.score, subcl.score, method="spearman")
cor.test(assoc.score, subcl.score, method="spearman", exact=FALSE)$p.value
plotCorrelation(assoc.score, subcl.score, 
    subtypes=mcols(gisticOV)$subtype, stcols=stcols)
```

## Assessing subclonality calls with PureCN

### Read and order
```{r pureCN.pp}
ppp.file <- file.path(data.dir, "PureCN_Purity_Ploidy.txt")
puriploi.pcn <- read.table(ppp.file) 
puriploi.pcn <- puriploi.pcn[!duplicated(puriploi.pcn[,1]),]
rownames(puriploi.pcn) <- puriploi.pcn[,"Sampleid"]
puriploi.pcn <- puriploi.pcn[,-1]
puriploi.pcn <- as.matrix(puriploi.pcn)
absdiff <- abs(puriploi.pcn[,c("purity", "ploidy")] - 
                puriploi.pcn[,c("Purity_wes", "Ploidy_wes")])
ind <- do.call(order, as.data.frame(absdiff))
puriploi.pcn <- puriploi.pcn[ind,]
head(puriploi.pcn)
cor(puriploi.pcn[,"purity"], puriploi.pcn[,"Purity_wes"], use="complete.obs")
cor(puriploi.pcn[,"Purity_wes"], puriploi.pcn[,"Purity_snp6"], use="complete.obs")
```

### Select large losses called subclonal by ABSOLUTE 
```{r selectAbsCalls}
selectAbsCalls <- function(sample, min.size=3000000)
{
    gr <- absGRL[[sample]]
    gr.3MB <- width(gr) > min.size
    is.loss <- (gr$Modal_HSCN_1 + gr$Modal_HSCN_2) < 2 
    is.subcl <- gr$score == 1

    ind <- gr.3MB & is.loss & is.subcl
    return(gr[ind])
}
(abs.calls <- selectAbsCalls(rownames(puriploi.pcn)[1]))
```

### liftOver hg19-based ABSOLUTE calls to compare with hg38-based PureCN calls 
- UCSC online liftOver (https://genome.ucsc.edu/cgi-bin/hgLiftOver)
- does smoothing and filtering on top of the chain mappings
- `rtracklayer::liftOver` only provides chain mappings

```{r liftOver}
chain.file <- file.path(data.dir, "hg19ToHg38.over.chain")
ch <- rtracklayer::import.chain(chain.file)
(hg19.call <- absGRL[[1]][1])
(blocks <- rtracklayer::liftOver(hg19.call, ch))
```

```{r liftOverFilterSmooth}
blocks <- unlist(blocks)
# harmonize chromosome by majority vote
tab <- table(as.character(seqnames(blocks))) 
nmax <- names(tab)[which.max(tab)]
# collapse blocks
blocks <- blocks[seqnames(blocks) == nmax]
(hg38.call <- range(reduce(blocks)))
```

### Select GISTIC regions of high subtype association and subclonality
```{r selecttopTables}
tests <- testSubtypes(gisticOV, ovsubs, what="full")
```

```{r selecttop}
# highest subtype association
ind <- order(assoc.score, decreasing=TRUE)
rowRanges(gisticOV)[ind[1:2]]
x <- tests[[ind[1]]]
x$observed
x$expected
diff <- x$observed - x$expected
(csums <- colSums(diff^2 / x$expected))
x <- tests[[ind[2]]]
x$observed
x$expected
diff <- x$observed - x$expected
(csums <- colSums(diff^2 / x$expected))

# highest subclonality
ind <- order(subcl.score, decreasing=TRUE)
rowRanges(gisticOV)[ind[1:2]]
x <- tests[[ind[1]]]
x$observed
x$expected
diff <- x$observed - x$expected
(csums <- colSums(diff^2 / x$expected))
x <- tests[[ind[2]]]
x$observed
x$expected
diff <- x$observed - x$expected
(csums <- colSums(diff^2 / x$expected))
```

# Other cancer types

```{r allCTypes}
allFile <- file.path(data.dir, "all_cancer_types.rds")
allCT <- readRDS(allFile)
```

## Number of samples
```{r plotNrSamples, fig.width=12, fig.height=6}
st <- sapply(allCT, function(ct) nrow(ct$subtypes))
gs <- sapply(allCT, function(ct) ncol(ct$gistic))
as <- sapply(allCT, function(ct) sum(rownames(ct$subtypes) %in% names(absGRL)))

plotNrSamples(gs, st, as) 
```

## Number of subtypes & GISTIC regions
```{r descrStats, fig.width=6, fig.height=6}
nr.sts <- sapply(allCT, function(ct) length(unique(ct$subtypes[,1])))
nr.gregs <- sapply(allCT, function(ct) nrow(ct$gistic))
plotNrSubtypesVsNrGisticRegions(nr.sts, nr.gregs)
```

## Subclonality
```{r subclDistrs, fig.width=12, fig.height=6}
subcl.scores <- lapply(allCT, function(ct) ct$subcl.score)
plotSubclonalityDistributions(subcl.scores)
```

## Correlation: subtype association & subclonality
```{r corVolcano, fig.width=9, fig.height=6}
rho <- sapply(allCT, function(ct) ct$rho)
p <- sapply(allCT, function(ct) ct$p)
volcanoCorrelation(rho, p)
```

## Intrinsic subtypes: sarcoma
```{r sarcSubtyAssoc}
gisticSARC <- gistic2RSE(ctype="SARC", peak="wide")
sarcsubs <- getBroadSubtypes(ctype="SARC", clust.alg="CNMF")
table(sarcsubs$cluster)
pvals <- testSubtypes(gisticSARC, sarcsubs, padj.method="none")
length(pvals)
adj.pvals <- p.adjust(pvals, method="BH")
sum(adj.pvals < 0.1)
hist(pvals, breaks=25, col="firebrick",xlab="Subtype association p-value", main="")
```


```{r sarcClinical}
sarc.clin <- RTCGAToolbox::getFirehoseData(dataset="SARC", runDate="20160128")
sarc.clin <- RTCGAToolbox::getData(sarc.clin, "clinical")
ids <- rownames(sarcsubs)
ids <- tolower(ids)
ids <- gsub("-", ".", ids)
ids <- substring(ids, 1, 12)
sarcsubs$histType <- sarc.clin[ids,"histological_type"] 
lapply(1:3, function(cl) table(sarcsubs[sarcsubs$cluster == cl,"histType"]))
```

```{r sarc}
assoc.score <- testSubtypes(gisticSARC, sarcsubs, what="statistic")
raSARC <- getMatchedAbsoluteCalls(absGRL, sarcsubs)
subcl.gisticSARC <- querySubclonality(raSARC, query=rowRanges(gisticSARC), sum.method="any")
subcl.score <- rowMeans(subcl.gisticSARC, na.rm=TRUE)
summary(subcl.score)
cor(assoc.score, subcl.score, method="spearman")
cor.test(assoc.score, subcl.score, method="spearman", exact=FALSE)$p.value
```

```{r sarcPlot, fig.width=6, fig.height=6}
sarc.cols <- stcols[c(1,3,2)]
names(sarc.cols) <- c("MFS/UPS", "LMS", "DDLPS")
sts <- testSubtypes(gisticSARC, sarcsubs, what="subtype")
sts <- names(sarc.cols)[sts]
plotCorrelation(assoc.score, subcl.score, subtypes=sts, stcols=sarc.cols)
```

<!--
## Intrinsic subtypes: breast cancer
Subtypes: luminal A, luminal B, basal-like, and HER2-enriched
```{r brcaSubtyAssoc}
gisticBRCA <- gistic2RSE(ctype="BRCA", peak="wide")
brcasubs <- getBroadSubtypes(ctype="BRCA", clust.alg="CNMF")
table(brcasubs$cluster)
pvals <- testSubtypes(gisticBRCA, brcasubs, padj.method="none")
length(pvals)
adj.pvals <- p.adjust(pvals, method="BH")
sum(adj.pvals < 0.1)
hist(pvals, breaks=25, col="firebrick",xlab="Subtype association p-value", main="")
```

```{r brca}
brca <- allCT$BRCA
names(brca)
brca[1:2]
```

## Late subtypes: glioblastoma
Subtypes: mesenchymal, proneural, neural, and classical
```{r gbmSubtyAssoc}
gisticGBM <- gistic2RSE(ctype="GBM", peak="wide")
gbmsubs <- getBroadSubtypes(ctype="GBM", clust.alg="CNMF")
table(gbmsubs$cluster)
pvals <- testSubtypes(gisticGBM, gbmsubs, padj.method="none")
length(pvals)
adj.pvals <- p.adjust(pvals, method="BH")
sum(adj.pvals < 0.1)
hist(pvals, breaks=25, col="firebrick",xlab="Subtype association p-value", main="")
```

```{r gbm}
gbm <- allCT$GBM
gbm[1:2]
```
-->

# Subtype classification in single-cell RNA-seq data 

```{r scRNAseq}
sc.file <- file.path(data.dir, "scRNAseq_logTPM.rds")
sc <- readRDS(sc.file)
sc[1:5, 1:5]
```

```{r explore.scRNAseq}
library(MASS)
data(GSE14764.eset)
boxplot(exprs(GSE14764.eset))
boxplot(sc)
ma.vals <- as.vector(exprs(GSE14764.eset))
sc.vals <- as.vector(sc)
truehist(ma.vals, nbins=50) 
truehist(sc.vals, nbins=50)
```

```{r exploreZeroInflat}
sc0 <- sc
for(i in seq_len(nrow(sc)))
    for(j in seq_len(ncol(sc)))
        if(is.na(sc[i,j])) sc0[i,j] <- 0
MASS::truehist(as.vector(sc0), nbins=50)
perc0 <- apply(sc0, 2, function(x) mean(x==0) * 100)
barplot(perc0,ylab="%genes with log2TPM = 0")
```
