---
title: "Subclonality of expression subtypes"
author: Ludwig Geistlinger, Sehyun Oh, Markus Riester, and Levi Waldron
output:
  BiocStyle::html_document:
    toc: true
    toc_depth: 2
vignette: >
  % \VignetteIndexEntry{Functional enrichment analysis of high-throughput omics data}
  % \VignetteEngine{knitr::rmarkdown}
---

```{r style, echo = FALSE}
suppressPackageStartupMessages({ 
    library(subtypeHeterogeneity) 
    library(consensusOV)
    library(RaggedExperiment)
    library(ComplexHeatmap)
    library(plotly)
})
```

# Setup
```{r setup}
library(subtypeHeterogeneity)
library(consensusOV)
library(RaggedExperiment)
library(ComplexHeatmap)
library(plotly)
```

```{r cols}
cb.pink <- "#CC79A7"
cb.red <- "#D55E00"
cb.blue <- "#0072B2"
cb.yellow <- "#F0E442"
cb.green <- "#009E73"
cb.lightblue <- "#56B4E9"
cb.orange <- "#E69F00"

stcols <- c(cb.lightblue, cb.green, cb.orange, cb.pink) 
names(stcols) <- c("PRO", "MES", "DIF", "IMR")
```

# Data sources

## Cancer types
```{r ctypes}
RTCGAToolbox::getFirehoseDatasets()
```

## ABSOLUTE
```{r absolute}
data.dir <- system.file("extdata", package="subtypeHeterogeneity") 
absFile <- file.path(data.dir, "ABSOLUTE_grangeslist.rds")
absGRL <- readRDS(absFile) 
absGRL
```

## GISTIC2
```{r gistic}
gisticOV <- gistic2RSE(ctype="OV", peak="wide")
gisticOV
rowRanges(gisticOV)
assay(gisticOV)[1:5,1:5]
```

## Expression-based subtypes
### Broad subtypes
```{r broadSubtys}
ovsubs <- getBroadSubtypes(ctype="OV", clust.alg="CNMF")
dim(ovsubs)
head(ovsubs)
table(ovsubs[,"cluster"])
```

### OV subtypes from different studies
```{r pooledSubtys}
pooled.file <- file.path(data.dir, "pooled_subtypes.rds")
pooled.subs <- readRDS(pooled.file)
table(pooled.subs[,"data.source"])
table(pooled.subs[,"Verhaak"])  
```

### TCGA subtype consistency
```{r subtyCons}
tab <- mapOVSubtypes(ovsubs, pooled.subs)
tab
(ind <- sort( apply(tab, 1, which.max) ))
```

```{r mapSubs}
sts <- names(ind)[ovsubs[,"cluster"]]
ovsubs <- data.frame(ovsubs, subtype=sts, stringsAsFactors=FALSE)
```

### Subtype purity & ploidy
```{r subtyPP}
pp.file <- file.path(data.dir, "ABSOLUTE_Purity_Ploidy.rds")
puri.ploi <- readRDS(pp.file)
head(puri.ploi)
```

```{r plotPP, fig.width=8, fig.height=8}
plotSubtypePurityPloidy(ovsubs, puri.ploi)
```

```{r stratPur}
sebin <- stratifyByPurity(ovsubs, puri.ploi, method="equal.bin")
lengths(sebin)
squint <- stratifyByPurity(ovsubs, puri.ploi, method="quintile")
lengths(squint)
```

```{r plotPurStrat, fig.width=5, fig.height=5}
plotPurityStrata(ovsubs, puri.ploi)
```

# Test: gistic-subtype association
```{r subtyAssoc}
pvals <- testSubtypes(gisticOV, ovsubs, padj.method="none")
adj.pvals <- p.adjust(pvals, method="BH")
length(adj.pvals)
head(adj.pvals)
sum(adj.pvals < 0.1)
```

```{r plotAssoc, fig.width=5, fig.height=5}
hist(pvals, breaks=25, col="firebrick",
        xlab="Subtype association p-value", main="")
```

## Genomic distribution of subtype-associated CNAs 

```{r tcgaOvCnvGenes}
cnv.genes <- getCnvGenesFromTCGA()
```

```{r circosSubAssoc, fig.width=10, fig.height=10}
sig.ind <- adj.pvals < 0.1
mcols(gisticOV)$subtype <- testSubtypes(gisticOV, ovsubs, what="subtype")
mcols(gisticOV)$significance <- ifelse(sig.ind, "*", "")
circosSubtypeAssociation(gisticOV, cnv.genes)
```

```{r cnaPerSubtype, fig.width=5, fig.height=5}
plotNrCNAsPerSubtype(mcols(gisticOV)$type[sig.ind], mcols(gisticOV)$subtype[sig.ind])
```

Annotate cytogenetic bands:

```{r tcgaOvCnvBands}
bands.file <- file.path(data.dir, "cytoBand_hg19.txt")
cbands <- read.delim(bands.file, header=FALSE)
cbands <- cbands[,-5]
colnames(cbands) <- c("seqnames", "start", "end", "band")
cbands[,4] <- paste0(sub("^chr", "", cbands[,1]), cbands[,4])
cbands <- makeGRangesFromDataFrame(cbands, keep.extra.columns=TRUE)
genome(cbands) <- "hg19"
cbands
gisticOV <- annotateCytoBands(gisticOV, cbands)
mcols(gisticOV)
```

```{r coocc, fig.width=7, fig.height=7}
coocc <- analyzeCooccurence(gisticOV) 
rownames(coocc) <- mcols(gisticOV)$band
ComplexHeatmap::Heatmap(coocc, show_row_names=TRUE, show_column_names=FALSE, 
    column_title="SCNAs", row_title="SCNAs", name="Co-occurrence", 
    row_names_gp = gpar(fontsize = 8))
```

# Subclonality

## Match subtyped samples and ABSOLUTE samples 
```{r matchCalls}
raOV <- getMatchedAbsoluteCalls(absGRL, ovsubs)
raOV
rowRanges(raOV)
```

## Summarize subclonality of ABSOLUTE calls in GISTIC regions 
```{r querySubcl}
subcl.gisticOV <- querySubclonality(raOV, query=rowRanges(gisticOV), sum.method="any")
dim(subcl.gisticOV)
subcl.gisticOV[1:5,1:5]
```

## Subclonality score
Def: Fraction of samples in which a mutation is subclonal.
```{r subclScore}
subcl.score <- rowMeans(subcl.gisticOV, na.rm=TRUE)
summary(subcl.score)
```

## Correlation: subtype-association score and subclonality score
```{r cor}
assoc.score <- testSubtypes(gisticOV, ovsubs, what="statistic")
cor(assoc.score, subcl.score, method="spearman")
cor.test(assoc.score, subcl.score, method="spearman", exact=FALSE)$p.value
#corPermTest(gisticOV, ovsubs, subcl.score)
```

```{r plotCor, fig.width=6, fig.height=6}
plotCorrelation(assoc.score, subcl.score) 
```

### Purity-stratified analysis
```{r purStrat}
sapply(squint, function(ids) analyzeStrata(ids, absGRL, gisticOV, ovsubs))
sapply(sebin[-1], function(ids) analyzeStrata(ids, absGRL, gisticOV, ovsubs))
```

### Only copy losses and single copy gains
```{r corNoHighAmp}
noHighAmp <- getMatchedAbsoluteCalls(absGRL, ovsubs, max.cn=3)
subcl.noHighAmp <- querySubclonality(noHighAmp, query=rowRanges(gisticOV))
subcl.noHighAmp <- rowMeans(subcl.noHighAmp, na.rm=TRUE)
cor(assoc.score, subcl.noHighAmp, method="spearman")
cor.test(assoc.score, subcl.noHighAmp, method="spearman", exact=FALSE)$p.value
#corPermTest(gisticOV, ovsubs, subcl.noHighAmp)
```

```{r plotCorNHA, fig.width=6, fig.height=6}
plotCorrelation(assoc.score, subcl.noHighAmp)
```

### Extend peaks by 500 kb up- and downstream 
```{r cor5kb, fig.width=6, fig.height=6}
subcl.gisticOV.5kb <- querySubclonality(raOV, query=rowRanges(gisticOV), ext=500000)
subcl.score <- rowMeans(subcl.gisticOV.5kb, na.rm=TRUE)
summary(subcl.score)
cor(assoc.score, subcl.score, method="spearman")
cor.test(assoc.score, subcl.score, method="spearman", exact=FALSE)$p.value
plotCorrelation(assoc.score, subcl.score, 
    subtypes=mcols(gisticOV)$subtype, stcols=stcols)
text(x=52.8, y=0.46, "20q11.21\n(BCL2L1)", pos=2)
text(x=50, y=0.52, "19p13.12\n(BRD4)", pos=2)
text(x=39.489, y=0.332, "12q15\n(FRS2)", pos=4)
text(x=32.232, y=0.492, "3q26.2\n(MECOM)", pos=3)
text(x=28.883, y=0.6, "8q24.21 (MYC)", pos=2)
text(x=29.801, y=0.563, "20q13.33", pos=2)
text(x=25.693, y=0.408, "2q31.2 (PDE11A)", pos=4)
text(x=26.991, y=0.353, "3p26.3", pos=4) 
text(x=17.59, y=0.272, "15q15.1 (MGA)", pos=4) 
text(x=12.389, y=0.22, "8p21.2 (PPP2R2A)", pos=4)
```

```{r annoOVranges}
rowRanges(gisticOV)$adj.pval <- adj.pvals
rowRanges(gisticOV)$subcl <- subcl.score 
```

## Assessing subclonality calls with PureCN

### Read and order
```{r pureCN.pp}
ppp.file <- file.path(data.dir, "PureCN_Purity_Ploidy.txt")
puriploi.pcn <- read.table(ppp.file) 
puriploi.pcn <- puriploi.pcn[!duplicated(puriploi.pcn[,1]),]
rownames(puriploi.pcn) <- puriploi.pcn[,"Sampleid"]
puriploi.pcn <- puriploi.pcn[,-1]
puriploi.pcn <- as.matrix(puriploi.pcn)
absdiff <- abs(puriploi.pcn[,c("purity", "ploidy")] - 
                puriploi.pcn[,c("Purity_wes", "Ploidy_wes")])
ind <- do.call(order, as.data.frame(absdiff))
puriploi.pcn <- puriploi.pcn[ind,]
head(puriploi.pcn)
cor(puriploi.pcn[,"purity"], puriploi.pcn[,"Purity_wes"], use="complete.obs")
cor(puriploi.pcn[,"Purity_wes"], puriploi.pcn[,"Purity_snp6"], use="complete.obs")
```

### CN concordance with ABSOLUTE
PureCN calls for S0293689 capture kit: 
```{r pureCN.calls}
pcn.call.file <- file.path(data.dir, "PureCN_OV_S0293689_grangeslist.rds")
pcn.calls <- readRDS(pcn.call.file)
pcn.calls
# select samples that intersect with ABSOLUTE data
isect.samples <- intersect(names(pcn.calls), names(absGRL))
pcn.calls <- pcn.calls[isect.samples]
pcn.calls.ra <- RaggedExperiment(pcn.calls) 
pcn.calls.ra
```

Get corresponding ABSOLUTE calls:
```{r isect.abs.calls}
abs.calls <- absGRL[isect.samples]
abs.calls <- as.data.frame(abs.calls)[,c(2:5,8:10)]
totalCN <- abs.calls[,"Modal_HSCN_1"] + abs.calls[,"Modal_HSCN_2"]
abs.calls <- data.frame(abs.calls[,1:4], CN=totalCN, score=abs.calls[,"score"])
abs.calls <- makeGRangesListFromDataFrame(abs.calls, 
    split.field="group_name", keep.extra.columns=TRUE)
abs.calls
abs.calls.ra <- RaggedExperiment(abs.calls)
abs.calls.ra <- abs.calls.ra[,colnames(pcn.calls.ra)]
```

OVC GISTIC2 regions on hg19 (ABSOLUTE) and hg38 (PureCN)
```{r gisticHG38}
gisticOV.ranges.hg19 <- file.path(data.dir, "gisticOV_ranges_hg19.txt") 
gisticOV.ranges.hg19 <- scan(gisticOV.ranges.hg19, what="character")
gisticOV.ranges.hg19 <- GRanges(gisticOV.ranges.hg19) 
gisticOV.ranges.hg19 
gisticOV.ranges.hg38 <- file.path(data.dir, "gisticOV_ranges_hg38.txt") 
gisticOV.ranges.hg38 <- scan(gisticOV.ranges.hg38, what="character")
gisticOV.ranges.hg38 <- GRanges(gisticOV.ranges.hg38) 
gisticOV.ranges.hg38
```

Summarize calls in GISTIC regions using either  
 
- the largest call overlapping a GISTIC region  
- a weighted mean of calls overlapping a GISTIC region 
```{r qred}
.largest <- function(score, range, qrange) 
{
    return.type <- class(score[[1]])
    default.value <- do.call(return.type, list(1))
    ind <- which.max(width(range))
    res <- vapply(seq_along(score), 
           function(i) score[[i]][ind[i]], default.value)
    return(res)
}
pcn.rassay <- qreduceAssay(pcn.calls.ra, query=gisticOV.ranges.hg38, 
                    simplifyReduce=.largest, background=2)
abs.rassay <- qreduceAssay(abs.calls.ra, query=gisticOV.ranges.hg19,
                    simplifyReduce=.largest, background=2)


.weightedmean <- function(score, range, qrange)
{
    w <- width(range)
    s <- sum(score * w) / sum(w)
    return(round(s))
}

pcn.rassay2 <- qreduceAssay(pcn.calls.ra, query=gisticOV.ranges.hg38,
                    simplifyReduce=.weightedmean, background=2)
abs.rassay2 <- qreduceAssay(abs.calls.ra, query=gisticOV.ranges.hg19,
                    simplifyReduce=.weightedmean, background=2)
```

Evaluate per-sample concordance:

- fraction of GISTIC regions with equal copy number  
- fraction of GISTIC regions agreeing in alteration type (amplification / deletion)  
- fraction of GISTIC regions that deviate by max. 1 copy

```{r evalConcord}
# largest
fract.equal <- vapply(seq_along(isect.samples), 
        function(i) mean(pcn.rassay[,i] == abs.rassay[,i]), 
        numeric(1))
fract.type <- vapply(seq_along(isect.samples), 
        function(i) mean(((pcn.rassay[,i] < 2) & (abs.rassay[,i] < 2)) |
                          ((pcn.rassay[,i] > 2) & (abs.rassay[,i] > 2)) |  
                          ((pcn.rassay[,i] == 2) & (abs.rassay[,i] == 2))), 
        numeric(1))
fract.diff1 <- vapply(seq_along(isect.samples), 
        function(i) mean((pcn.rassay[,i] == abs.rassay[,i]) | 
                            (pcn.rassay[,i] == abs.rassay[,i] + 1) |
                            (pcn.rassay[,i] == abs.rassay[,i] - 1)), 
        numeric(1))

# weighted mean
fract.equal2 <- vapply(seq_along(isect.samples), 
        function(i) mean(pcn.rassay2[,i] == abs.rassay2[,i]), 
        numeric(1))
fract.type2 <- vapply(seq_along(isect.samples), 
        function(i) mean(((pcn.rassay2[,i] < 2) & (abs.rassay2[,i] < 2)) |
                          ((pcn.rassay2[,i] > 2) & (abs.rassay2[,i] > 2)) |  
                          ((pcn.rassay2[,i] == 2) & (abs.rassay2[,i] == 2))), 
        numeric(1))
fract.diff12 <- vapply(seq_along(isect.samples), 
        function(i) mean((pcn.rassay2[,i] == abs.rassay2[,i]) | 
                            (pcn.rassay2[,i] == abs.rassay2[,i] + 1) |
                            (pcn.rassay2[,i] == abs.rassay2[,i] - 1)), 
        numeric(1))
```

Plot concordance:
```{r plotConcord}
par(mar=c(5.1, 4.1, 4.1, 8.1), xpd=TRUE)
boxplot(fract.equal, fract.equal2, 
        fract.type, fract.type2,
        fract.diff1, fract.diff12, col=rep(c(cb.blue, cb.red), 3),
        names=rep(c("equal", "type", "diff1"), each=2),
        ylab="Fraction of concordant GISTIC2 regions")
legend("topright", inset=c(-0.3,0), legend=c("largest", "wmean"),
        col=c(cb.blue, cb.red), lwd=2)
```


### Subclonality concordance with ABSOLUTE 

Summarize subclonality of GISTIC2 regions

```{r subclConcord}
# PureCN
maxScore <- function(score, range, qrange) max(score, na.rm=TRUE)
pcn.rassay.sc <- RaggedExperiment::qreduceAssay(pcn.calls.ra, 
                query=gisticOV.ranges.hg38, 
                simplifyReduce=maxScore, i="score", background=NA)
pcn.subcl <- rowMeans(pcn.rassay.sc, na.rm=TRUE)
summary(pcn.subcl)

#ABSOLUTE
abs.rassay.sc <- RaggedExperiment::qreduceAssay(abs.calls.ra, 
                query=gisticOV.ranges.hg19, 
                simplifyReduce=maxScore, i="score", background=NA)
abs.subcl <- rowMeans(abs.rassay.sc, na.rm=TRUE)
summary(abs.subcl)
```

Correlation 
```{r corConcord}
cor.test(pcn.subcl, abs.subcl, method="spearman")
```


```{r selRegions}
sort(abs.subcl[pcn.subcl > 0.4], decreasing=TRUE)
```

### Select large losses called subclonal by ABSOLUTE 
```{r selectAbsCalls}
selectAbsCalls <- function(sample, min.size=3000000)
{
    gr <- absGRL[[sample]]
    gr.3MB <- width(gr) > min.size
    is.loss <- (gr$Modal_HSCN_1 + gr$Modal_HSCN_2) < 2 
    is.subcl <- gr$score == 1

    ind <- gr.3MB & is.loss & is.subcl
    return(gr[ind])
}
(abs.calls <- selectAbsCalls(rownames(puriploi.pcn)[1]))
```

### liftOver hg19-based ABSOLUTE calls to compare with hg38-based PureCN calls 
- UCSC online liftOver (https://genome.ucsc.edu/cgi-bin/hgLiftOver)
- does smoothing and filtering on top of the chain mappings
- `rtracklayer::liftOver` only provides chain mappings

```{r liftOver}
chain.file <- file.path(data.dir, "hg19ToHg38.over.chain")
ch <- rtracklayer::import.chain(chain.file)
(hg19.call <- absGRL[[1]][1])
(blocks <- rtracklayer::liftOver(hg19.call, ch))
```

```{r liftOverFilterSmooth}
blocks <- unlist(blocks)
# harmonize chromosome by majority vote
tab <- table(as.character(seqnames(blocks))) 
nmax <- names(tab)[which.max(tab)]
# collapse blocks
blocks <- blocks[seqnames(blocks) == nmax]
(hg38.call <- range(reduce(blocks)))
```

### Select GISTIC regions of high subtype association and subclonality
```{r selecttopTables}
tests <- testSubtypes(gisticOV, ovsubs, what="full")
extraL <- stratifySubclonality(gisticOV, subcl.gisticOV.5kb, ovsubs, tests)
```

```{r selecttop, fig.width=6, fig.height=6}
# highest subtype association
ind <- order(assoc.score, decreasing=TRUE)
rowRanges(gisticOV)[ind[1:2]]
# 20q11.21 (BCL2L1)
x <- tests[[ind[1]]]
(ot <- x$observed)
x$expected
diff <- x$observed - x$expected
(csums <- colSums(diff^2 / x$expected))
# 19p13.12 (BRD4)
x <- tests[[ind[2]]]
(ot <- x$observed)
x$expected
diff <- x$observed - x$expected
(csums <- colSums(diff^2 / x$expected))

# highest subclonality
ind <- order(subcl.score, decreasing=TRUE)
rowRanges(gisticOV)[ind[1:2]]
```

```{r selecttopplotOV, fig.width=6, fig.height=6}
# highest subtype association
ind <- order(assoc.score, decreasing=TRUE)
top.genes <- c("BCL2L1", "BRD4", "FRS2", "MECOM", "ID4", "", "", "MYC", "", "") 
for(i in 1:10)
{
    x <- ind[i]
    type <- rowRanges(gisticOV)$type[x]
    type <- ifelse(type == "Deletion", "loss", "gain") 
    bpSubtypeStrata(extraL[[x]], type=type)
    title(main=paste0(rowRanges(gisticOV)$band[x], " (", top.genes[i], ")"))
}
```


Plot regions:
```{r intersectSamplesABSOLUTEandSubtype}
isamples <- intersect(rownames(ovsubs), names(absGRL))
absov <- absGRL[isamples]
(ov.rel <- table(ovsubs[names(absov),"cluster"]))
```

```{r plotHighSubcl, fig.width=6, fig.height=6}
# myc
r <- GRanges("chr8:128694547-129309687")
g <- GRanges(c("chr8:128747680-128753674", "chr8:128806779-129113499"))
names(g) <- c("MYC", "PVT1")
genome(g) <- genome(r) <- "hg19" 

rb <- GRanges("chr13:48875329-49064807")

ov.olaps <- subsetByOverlaps(unlist(absov), r)
#as a GRangesList: splitAsList(olaps, names(olaps))
ov.olaps$State <- ovsubs[names(ov.olaps), "cluster"]
gvizRegion(r, g, ov.olaps, ov.rel=ov.rel, type="+")
```


# Other cancer types

```{r allCTypes}
allFile <- file.path(data.dir, "all_cancer_types.rds")
allCT <- readRDS(allFile)
```

## Number of samples
```{r plotNrSamples, fig.width=12, fig.height=6}
st <- sapply(allCT, function(ct) nrow(ct$subtypes))
gs <- sapply(allCT, function(ct) ncol(ct$gistic))
as <- sapply(allCT, function(ct) sum(rownames(ct$subtypes) %in% names(absGRL)))

plotNrSamples(gs, st, as) 
```

## Number of subtypes & GISTIC regions
```{r descrStats, fig.width=6, fig.height=6}
nr.sts <- sapply(allCT, function(ct) length(unique(ct$subtypes[,1])))
nr.gregs <- sapply(allCT, function(ct) nrow(ct$gistic))
plotNrSubtypesVsNrGisticRegions(nr.sts, nr.gregs)
```

## Subclonality
```{r subclDistrs, fig.width=12, fig.height=6}
subcl.scores <- lapply(allCT, function(ct) ct$subcl.score)
plotSubclonalityDistributions(subcl.scores)
```

## Correlation: subtype association & subclonality
```{r corVolcano, fig.width=9, fig.height=6}
rho <- sapply(allCT, function(ct) ct$rho)
p <- sapply(allCT, function(ct) ct$p)
volcanoCorrelation(rho, p)
```

## Intrinsic subtypes: sarcoma
```{r sarcSubtyAssoc}
gisticSARC <- gistic2RSE(ctype="SARC", peak="wide")
sarcsubs <- getBroadSubtypes(ctype="SARC", clust.alg="CNMF")
table(sarcsubs$cluster)
pvals <- testSubtypes(gisticSARC, sarcsubs, padj.method="none")
length(pvals)
adj.pvals <- p.adjust(pvals, method="BH")
sum(adj.pvals < 0.1)
hist(pvals, breaks=25, col="firebrick",xlab="Subtype association p-value", main="")
```


```{r sarcClinical}
sarc.clin <- RTCGAToolbox::getFirehoseData(dataset="SARC", runDate="20160128")
sarc.clin <- RTCGAToolbox::getData(sarc.clin, "clinical")
ids <- rownames(sarcsubs)
ids <- tolower(ids)
ids <- gsub("-", ".", ids)
ids <- substring(ids, 1, 12)
sarcsubs$histType <- sarc.clin[ids,"histological_type"] 
lapply(1:3, function(cl) table(sarcsubs[sarcsubs$cluster == cl,"histType"]))
```

```{r sarc}
assoc.score <- testSubtypes(gisticSARC, sarcsubs, what="statistic")
raSARC <- getMatchedAbsoluteCalls(absGRL, sarcsubs)
subcl.gisticSARC <- querySubclonality(raSARC, query=rowRanges(gisticSARC), sum.method="any")
subcl.score <- rowMeans(subcl.gisticSARC, na.rm=TRUE)
summary(subcl.score)
cor(assoc.score, subcl.score, method="spearman")
cor.test(assoc.score, subcl.score, method="spearman", exact=FALSE)$p.value
```

```{r sarcPlot, fig.width=6, fig.height=6}
sarc.cols <- stcols[c(1,3,2)]
names(sarc.cols) <- c("MFS/UPS", "LMS", "DDLPS")
sts <- testSubtypes(gisticSARC, sarcsubs, what="subtype")
sts <- names(sarc.cols)[sts]
plotCorrelation(assoc.score, subcl.score, subtypes=sts, stcols=sarc.cols, lpos="right")
# top assoc
text(x=56.885, y=0.548, "1p36.32\n(TP73)", pos=2)
text(x=36.908, y=0.336, "6q25.1\n(UST)", pos=4)
text(x=33.943, y=0.275, "10q23.31 (PTEN)", pos=2)
text(x=30.257, y=0.249, "13q14.2 (RB1)", pos=2)
text(x=27.44, y=0.435, "10q26.3\n(SPRN)", pos=4)
text(x=25.501, y=0.362, "1p32.1\n(JUN)", pos=4)
# top subcl
text(x=10.116, y=0.679, "8q", pos=4)
text(x=6.37, y=0.678, "9q34", pos=2)
text(x=8.586, y=0.662, "17q", pos=4)
text(x=7.658, y=0.65, "5p15.33\n(TERT)", pos=2)
text(x=19.497, y=0.625, "8p23.2 (CSMD1)", pos=4)
text(x=17.4, y=0.625, "20q13.33", pos=2)
text(x=55.915, y=0.262, "12q15 (MDM2)", pos=2)
```

```{r annoSARCranges}
gisticSARC <- annotateCytoBands(gisticSARC, cbands)
rowRanges(gisticSARC)$assoc.score <- assoc.score
rowRanges(gisticSARC)$subcl <- subcl.score 
rowRanges(gisticSARC)$subtype <- sts 
ind.assoc <- order(assoc.score, decreasing=TRUE)
ind.subcl <- order(subcl.score, decreasing=TRUE)
rowRanges(gisticSARC)[ind.assoc]
rowRanges(gisticSARC)[ind.subcl]
```

```{r selecttopSARC}
tests <- testSubtypes(gisticSARC, sarcsubs, what="full")
extraL <- stratifySubclonality(gisticSARC, subcl.gisticSARC, 
    sarcsubs, tests, st.names=names(sarc.cols))
```

```{r selecttopplot, fig.width=6, fig.height=6}
# highest subtype association
ind <- ind.assoc
sarc.genes <- c("TP73", "MDM2", "UST", "PTEN", "RB1", "SPRN", "JUN", "", "NF1") 
for(i in 1:9)
{
    x <- ind[i]
    type <- rowRanges(gisticSARC)$type[x]
    type <- ifelse(type == "Deletion", "loss", "gain") 
    bpSubtypeStrata(extraL[[x]], type=type)
    title(main=paste0(rowRanges(gisticSARC)$band[x], " (", sarc.genes[i], ")"))
}
```

```{r histcluster}
histcl <- rep(4, nrow(sarcsubs))
ind <- grep("myxofibrosarcoma", sarcsubs$histType)
histcl[ind] <- 1
ind <- grep("undifferentiated pleomorphic sarcoma", sarcsubs$histType)
histcl[ind] <- 1
ind <- grep("leiomyosarcoma", sarcsubs$histType)
histcl[ind] <- 2
ind <- grep("dedifferentiated liposarcoma", sarcsubs$histType)
histcl[ind] <- 3
histsubs <- sarcsubs[histcl != 4,]
histsubs$cluster <- histcl[histcl !=4]  
```

```{r testHistSubs}
assoc.score <- testSubtypes(gisticSARC, histsubs, what="statistic")
raSARC <- getMatchedAbsoluteCalls(absGRL, histsubs)
subcl.gisticSARC <- querySubclonality(raSARC, query=rowRanges(gisticSARC), sum.method="any")
subcl.score <- rowMeans(subcl.gisticSARC, na.rm=TRUE)
summary(subcl.score)
cor(assoc.score, subcl.score, method="spearman")
cor.test(assoc.score, subcl.score, method="spearman", exact=FALSE)$p.value
```

### Overlapping SCNAs between OV and SARC
```{r ovsarcranges}
sarcranges <- rowRanges(gisticSARC)
sarcranges$adj.pval <- adj.pvals
sarcranges$subcl <- subcl.score
ovranges <- rowRanges(gisticOV)
plotCommonSCNAs(ovranges, sarcranges)
```

Plot regions:
```{r intersectSamplesABSOLUTEandSubtypeSARC}
isamples <- intersect(rownames(sarcsubs), names(absGRL))
abssarc <- absGRL[isamples]
```

```{r plotHighSubclSarc}
# myc
sarc.olaps <- subsetByOverlaps(unlist(abssarc), r)
#as a GRangesList: splitAsList(olaps, names(olaps))
sarc.olaps$State <- sarcsubs[names(sarc.olaps), "cluster"]
gvizRegion(r, g, ov.olaps, sarc.olaps)
```


<!--
## Intrinsic subtypes: breast cancer
Subtypes: luminal A, luminal B, basal-like, and HER2-enriched
```{r brcaSubtyAssoc}
gisticBRCA <- gistic2RSE(ctype="BRCA", peak="wide")
brcasubs <- getBroadSubtypes(ctype="BRCA", clust.alg="CNMF")
table(brcasubs$cluster)
pvals <- testSubtypes(gisticBRCA, brcasubs, padj.method="none")
length(pvals)
adj.pvals <- p.adjust(pvals, method="BH")
sum(adj.pvals < 0.1)
hist(pvals, breaks=25, col="firebrick",xlab="Subtype association p-value", main="")
```

```{r brca}
brca <- allCT$BRCA
names(brca)
brca[1:2]
```

## Late subtypes: glioblastoma
Subtypes: mesenchymal, proneural, neural, and classical
```{r gbmSubtyAssoc}
gisticGBM <- gistic2RSE(ctype="GBM", peak="wide")
gbmsubs <- getBroadSubtypes(ctype="GBM", clust.alg="CNMF")
table(gbmsubs$cluster)
pvals <- testSubtypes(gisticGBM, gbmsubs, padj.method="none")
length(pvals)
adj.pvals <- p.adjust(pvals, method="BH")
sum(adj.pvals < 0.1)
hist(pvals, breaks=25, col="firebrick",xlab="Subtype association p-value", main="")
```

```{r gbm}
gbm <- allCT$GBM
gbm[1:2]
```
-->

# Subtype classification in single-cell RNA-seq data 

## Exploratory analysis
```{r scRNAseq}
sc.file <- file.path(data.dir, "scRNAseq_logTPM.rds")
sc <- readRDS(sc.file)
sc[1:5, 1:5]
```

```{r explore.scRNAseq}
library(MASS)
data(GSE14764.eset)
boxplot(exprs(GSE14764.eset))
boxplot(sc)
ma.vals <- as.vector(exprs(GSE14764.eset))
sc.vals <- as.vector(sc)
truehist(ma.vals, nbins=50) 
truehist(sc.vals, nbins=50)
```

```{r exploreZeroInflat}
sc0 <- sc
for(i in seq_len(nrow(sc)))
    for(j in seq_len(ncol(sc)))
        if(is.na(sc[i,j])) sc0[i,j] <- 0
MASS::truehist(as.vector(sc0), nbins=50)
perc0 <- apply(sc0, 2, function(x) mean(x==0) * 100)
barplot(perc0,ylab="%genes with log2TPM = 0")
```

## Subtype classification

Genes considered by the `consensusOV` classifier:
```{r consensusGenes}
library(consensusOV)
fnames <- lapply(
    consensusOV:::esets.rescaled.classified.filteredgenes, 
    Biobase::featureNames)
clgenes <- sort(unique(unlist(fnames)))
clgenes <- sub("^geneid.", "", clgenes)
clgenes
```

Single cell data restricted to classifier genes:
```{r scRestricted}
sc.file <- file.path(data.dir, "scRNAseq_consensusOV_genes.txt")
sc.expr <- as.matrix(read.delim(sc.file, row.names=1L))
dim(sc.expr)
sc.expr[1:5,1:5]
# exclude all zero 
ind <- apply(sc.expr, 1, function(x) all(x == 0))
sc.expr <- sc.expr[!ind,]
dim(sc.expr)
```

Get consensus subtypes:
```{r consensusSubtypes}
# consensus.subtypes <- get.consensus.subtypes(sc.expr, rownames(sc.expr))
res.file <- file.path(data.dir, "consensus_subtypes_verhaak100.rds")
consensus.subtypes <- readRDS(res.file)
```

Cell type annotation:
```{r cellAnnotation}
c2t.file <- file.path(data.dir, "scRNAseq_celltypes.txt")
c2t <- read.delim(c2t.file, as.is=TRUE)
n <- paste0("X", c2t[,1])
c2t <- c2t[,2]
names(c2t) <- n
head(c2t) 
table(c2t)
```

Expression heatmap incl. subtype and cell type annotation:
```{r scHeatmap}
sind <- 2:ncol(sc.expr)
sc.subtypes <- consensus.subtypes
sc.subtypes$consensusOV.subtypes <- sc.subtypes$consensusOV.subtypes[sind]
sc.subtypes$rf.probs <- sc.subtypes$rf.probs[sind, c(1:2,4,3)]
scHeatmap(sc.expr[,sind], sc.subtypes)
```

Subtype by cell type:
```{r stByCT}
# subtype
st <- consensus.subtypes$consensusOV.subtypes[sind]
st <- sub("_consensus$", "", st)
ct <- c2t[colnames(sc.expr)[sind]]

ept <- table(st[ct=="Epithelial"])
ept <- round(ept / sum(ept), digits=3)
ept
stro <- table(st[ct=="Stroma"])
stro <- round(stro / sum(stro), digits=3)
stro
```

Pie chart of subtype classification:
```{r plotlyPie, eval=FALSE}
ept <- ept[ept > 0]
cind <- match(names(ept), names(stcols))
plotlyPie(labels=names(ept), values=ept, colors=stcols[cind])
```

Display classification probabilites:
```{r rfprobHeatmap}
dat <- consensus.subtypes$rf.probs[sind,]
colnames(dat) <- sub("_consensus$", "", colnames(dat))
library(ComplexHeatmap)
Heatmap(dat, show_row_names=FALSE, row_title="Cells", name="Prob")
```

Consider only top50% confident classifications: 
```{r top50}
my.predictions.margins <- subtypeHeterogeneity::margin(consensus.subtypes$rf.probs)
# bulk 
my.predictions.margins[1]
# single cells
my.predictions.margins <- my.predictions.margins[sind]
ma95 <- quantile(my.predictions.margins, 0.95)
summary(my.predictions.margins)
# top50
ecdf.evaluated <- ecdf(my.predictions.margins)(my.predictions.margins)
na.ind <- ecdf.evaluated < 0.5
my.predictions.margins[na.ind] <- NA
top50 <- !is.na(my.predictions.margins)
```

Max prob (subtype call) and margin score (diff to second highest prob) distribution:
```{r marginDist}
maxp <- Biobase::rowMax(consensus.subtypes$rf.probs[sind,])[top50]
marg <- my.predictions.margins[top50]
boxplot(maxp, marg, names=c("max", "margin"), ylab="class.prob")
```

```{r stByCT-top50}
ept <- table(st[top50][ct[top50]=="Epithelial"])
ept <- round(ept / sum(ept), digits=3)
ept
stro <- table(st[top50][ct[top50]=="Stroma"])
stro <- round(stro / sum(stro), digits=3)
stro
```

Repeat analysis for logTPM > 3 (UofW single-cell threshold)
```{r exclLowExpr, eval=FALSE}
expr <- log(sc.expr + 1, base=2)
ind <- rowSums(expr) > 3
consensus.subtypes <- get.consensus.subtypes(
    sc.expr[ind,], 
    rownames(sc.expr)[ind])
```

Margin scores in comparison to TCGA bulk: 
```{r marginCompTCGA}
m100 <- c(0.0060, 0.1425, 0.2450, 0.2551, 0.3580, 0.8460)
m800 <- c(0.0020, 0.1270, 0.2050, 0.2061, 0.2860, 0.4420)
tcga.ma <- c(0.0060,  0.2265,  0.5250,  0.5196,  0.8115,  1.0000)
tcga.rseq <- c(0.0000,  0.2700,  0.6020,  0.5524,  0.8280,  1.0000)
names(m100) <- c("Min.", "1st Qu.",  "Median", "Mean", "3rd Qu.",    "Max.") 
names(m800) <- names(tcga.ma) <- names(tcga.rseq) <- names(m100)    

# base plot
par(pch=20)
par(las=1)
boxplot(m100, m800, tcga.rseq, tcga.ma, ylab="margin score", 
    names=c("scRNAseq100", "scRNAseq800", "TCGA-RNAseq", "TCGA-Marray"), 
    col=c(rep(cb.red,2), rep(cb.blue,2)), varwidth=TRUE)
points(c(1,2), y=c(0.42, 0.452), col=cb.red)
text(c(1,2), y=c(0.42, 0.452), "Bulk", col=cb.red, pos=3)

# ggplot
df <- data.frame(
    x=c("scRNAseq100", "scRNAseq800", "TCGA-RNAseq", "TCGA-Marray"), 
    t(cbind(m100, m800, tcga.rseq, tcga.ma)))
colnames(df)[2:7] <- c("y0", "y25", "y50", "mean", "y75", "y100") 
df[,1] <- factor(df[,1], levels=rev(df[,1]))
df[1,"y100"] <- ma95

df.points <- data.frame(x=c(4,3,4), y=c(0.42, 0.452, 0.846))

ggplot() + 
    geom_boxplot(data=df,
        aes(x=x, ymin=y0, lower=y25, middle=y50, upper=y75, ymax=y100), 
        stat="identity", fill=factor(rep(c(cb.blue, cb.red), each=2))) + 
    geom_point(data=df.points[1:2,], aes(x=x, y=y), color=cb.red) +
    geom_point(data=df.points[3,], aes(x=x, y=y)) +
    geom_text(data=df.points[1:2,], aes(x=x, y=y), 
                color=cb.red, label="Bulk", hjust=0, nudge_x = 0.2) +
    xlab("") + ylab("margin score") + coord_flip()

```

## Extension of consensus classifier

```{r getExtVerhaakSig}
# Supplementary Table S8A from Verhaak et al, JCI, 2013
verhaak.file <- file.path(data.dir, "Verhaak_supplementT8A.txt")
verhaak800 <- getExtendedVerhaakSignature(verhaak.file, nr.genes.per.subtype=200)
length(verhaak800)
head(verhaak800)
```

```{r geVerhaakSig}
# Supplementary Table S7 from Verhaak et al, JCI, 2013
verhaak100.file <- file.path(data.dir, "Verhaak_100genes.txt")
verhaak100 <- read.delim(verhaak100.file, as.is=TRUE)
verhaak100 <- verhaak100[2:101,1]
```

Interesting:
```{r ovlpVerhaak100-800}
table(verhaak100 %in% verhaak800)
```

Also interesting:
```{r ovlpVerhaakSigConsensusSig}
consensus.sig <- fData(consensusOV:::consensus.training.dataset.full)$gene
table(consensus.sig %in% verhaak100)
```

Merge:
```{r mergeVerhaak100-800}
verhaak800 <- sort(unique(c(verhaak100, verhaak800)))
```

Build extended training set based on extended verhaak signature:
```{r buildExtTraining}
selectGenes <- function(eset)
{
    ind <- fData(eset)$gene %in% verhaak800
    return(eset[ind,]) 
}
eset.file <- file.path(data.dir, "esets.rescaled.classified.filtered.RData")
esets <- get(load(eset.file))
esets.merged <- consensusOV:::dataset.merging(esets, method = "intersect", standardization = "none")
dim(esets.merged)
# sanity check
fnames <- featureNames(consensusOV:::consensus.training.dataset.full)[2:3]
snames <- sampleNames(consensusOV:::consensus.training.dataset.full)[1:5]
exprs(esets.merged)[fnames, snames]
exprs(consensusOV:::consensus.training.dataset.full)[2:3,1:5]
training.ext <- esets.merged
```

Single cell data restricted to verhaak800 signatures:
```{r scExtRestricted}
sc.file <- file.path(data.dir, "scRNAseq_verhaak800.txt")
sc.expr <- as.matrix(read.delim(sc.file, row.names=1L))
dim(sc.expr)
sc.expr[1:5,1:5]
# exclude NA
ind <- apply(sc.expr, 1, function(x) any(is.na(x)))
sc.expr <- sc.expr[!ind,]
# exclude all zero 
ind <- apply(sc.expr, 1, function(x) all(x == 0))
sc.expr <- sc.expr[!ind,]
dim(sc.expr)
```


```{r consSubtypesExt}
# takes a while
# consensus.subtypes <- get.consensus.subtypes(sc.expr, rownames(sc.expr), .training.dataset=training.ext)
res.file <- file.path(data.dir, "consensus_subtypes_verhaak800.rds")
consensus.subtypes <- readRDS(res.file)
table(consensus.subtypes$consensusOV.subtypes)
head(consensus.subtypes$rf.probs)
# Margins
my.predictions.margins <- subtypeHeterogeneity::margin(consensus.subtypes$rf.probs)
# Bulk
my.predictions.margins[1]
# Cells
summary(my.predictions.margins[sind])
sc.subtypes <- consensus.subtypes
sc.subtypes$consensusOV.subtypes <- sc.subtypes$consensusOV.subtypes[sind]
sc.subtypes$rf.probs <- sc.subtypes$rf.probs[sind, c(1:2,4,3)]
scHeatmap(sc.expr[,sind], sc.subtypes)
```

```{r consSubtypesExt2}
expr <- log(sc.expr + 1, base=2)
ind <- rowSums(expr) > 3
dim(sc.expr[ind,])
# consensus.subtypes2 <- get.consensus.subtypes(sc.expr[ind,], rownames(sc.expr)[ind], .training.dataset=esets.merged)
res.file <- file.path(data.dir, "consensus_subtypes_verhaak800_logTPMgr3.rds")
consensus.subtypes2 <- readRDS(res.file)
table(consensus.subtypes2$consensusOV.subtypes)
head(consensus.subtypes2$rf.probs)
# Margins
my.predictions.margins <- subtypeHeterogeneity::margin(consensus.subtypes2$rf.probs)
# Bulk
my.predictions.margins[1]
# Cells
summary(my.predictions.margins[sind])
```

